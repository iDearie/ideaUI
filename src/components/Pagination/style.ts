import { StyleSheet, ViewStyle, TextStyle } from 'react-native';
import { Theme } from '../Theme';

export interface PaginationStyles {
  container: ViewStyle;
  number: ViewStyle;
  current: TextStyle;
  btn: ViewStyle;
}

export const styles = (theme: Theme) =>
  StyleSheet.create<PaginationStyles>({
    container: {
      flexDirection: 'row',
      alignItems: 'center',
      justifyContent: 'center'
    },
    number: {
      flex: 1,
      alignItems: 'center'
    },
    current: {
      color: theme.main_color
    },
    btn: {
      paddingHorizontal: 10
    }
  });
