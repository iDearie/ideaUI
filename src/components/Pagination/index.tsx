import PropTypes from 'prop-types';
import React, { Component } from 'react';
import { Text, View } from 'react-native';
import { BaseProps } from '../base/Props';
import { Button } from '../Button';
import { WithTheme } from '../Theme';
import { styles } from './style';
import { getComponentLocale } from '../_util/getLocale';
import zh_CN from './locale/zh_CN';

interface PaginationProps extends BaseProps {
  current: number; //当前页号
  total: number; //数据总数
  simple?: boolean; //是否隐藏数值
  disabled?: boolean; //禁用状态
  locale?: { prevText: string; nextText: string };
  onChange?: (e: { current: number }) => void;
}

interface PaginationState {
  current: number;
}

class Pagination extends Component<PaginationProps, PaginationState> {
  static defaultProps: Partial<PaginationProps> = {
    current: 2,
    total: 5,
    simple: false,
    disabled: false
  };

  static contextTypes = {
    contextLocale: PropTypes.object
  };

  constructor(props: PaginationProps) {
    super(props);
    this.state = {
      current: props.current
    };
  }

  componentWillReceiveProps(nextProps: PaginationProps) {
    const { current } = nextProps;
    this.setState({
      current
    });
  }

  onPressPrev = () => {
    const { current } = this.state;
    const { onChange } = this.props;
    this.setState(() => {
      const _current = !(current - 1) ? current : current - 1;
      onChange && onChange({ current: _current });
      return {
        current: _current
      };
    });
  };

  onPressNext = () => {
    const { current } = this.state;
    const { total, onChange } = this.props;
    this.setState(() => {
      const _current = current === total ? current : current + 1;
      onChange && onChange({ current: _current });
      return {
        current: _current
      };
    });
  };

  render() {
    const { total, simple, disabled } = this.props;
    const { current } = this.state;
    const locale = getComponentLocale(this.props, (this as any).context, 'Pagination', () => zh_CN);
    const { prevText, nextText } = locale;
    return (
      <WithTheme themeStyles={styles}>
        {(_style) => (
          <View style={[_style.container]}>
            <Button
              disabled={disabled || current === 1}
              onPress={this.onPressPrev}
              type={'ghost'}
              size={'small'}
              viewStyle={[_style.btn]}>
              {prevText}
            </Button>
            <View style={[_style.number]}>
              {!simple ? (
                <Text>
                  <Text style={[_style.current]}>{current}</Text> / {total}
                </Text>
              ) : null}
            </View>
            <Button
              onPress={this.onPressNext}
              disabled={disabled || current === total}
              type={'ghost'}
              size={'small'}
              viewStyle={[_style.btn]}>
              {nextText}
            </Button>
          </View>
        )}
      </WithTheme>
    );
  }
}

export default Pagination;
