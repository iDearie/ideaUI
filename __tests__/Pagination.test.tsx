import React from 'react';
import renderer from 'react-test-renderer';
import Pagination from '../src/components/Pagination';
import { ReactWrapper, mount } from 'enzyme';

describe('Pagination', () => {
  it('组件快照', () => {
    const renderedValue = renderer.create(<Pagination current={1} total={20} />).toJSON();
    expect(renderedValue).toMatchSnapshot();
  });

  describe('rendering', () => {
    let wrapper: ReactWrapper;

    beforeEach(() => {
      wrapper = mount(<Pagination current={1} total={20} />);
    });

    it('按钮渲染', () => {
      const buttonList = (prevText: string = '上一页', nextText: string = '下一页') =>
        wrapper.findWhere((node) => {
          const type: string = node.name();
          const text: string = node.text();
          return type === 'Button' && (text === prevText || text === nextText);
        });
      expect(buttonList()).toHaveLength(2);
      wrapper.setProps({ locale: { prevText: '前', nextText: '后' } });
      expect(buttonList('前', '后')).toHaveLength(2);
    });

    it('simple模式渲染', () => {
      wrapper.setProps({ simple: true });
      const simpleWrap = wrapper.findWhere((node) => {
        return node.name() === 'Text' && node.text() === '1 / 20';
      });
      expect(simpleWrap).toHaveLength(0);
    });

    it('按钮事件', () => {
      const prevButton = wrapper.findWhere((node) => node.name() === 'Button' && node.text() === '上一页');
      const nextButton = wrapper.findWhere((node) => node.name() === 'Button' && node.text() === '下一页');
      const prevButtonProps: any = prevButton.props();
      const nextButtonProps: any = nextButton.props();
      expect(prevButtonProps.disabled).toBe(true);
      expect(nextButtonProps.disabled).toBe(false);
      prevButtonProps.onPress();
      expect(wrapper.state('current')).toBe(1);
      nextButtonProps.onPress();
      expect(wrapper.state('current')).toBe(2);
      wrapper.setProps({ current: 20 });
      nextButtonProps.onPress();
      expect(wrapper.state('current')).toBe(20);
      prevButtonProps.onPress();
      expect(wrapper.state('current')).toBe(19);
    });
  });
});
