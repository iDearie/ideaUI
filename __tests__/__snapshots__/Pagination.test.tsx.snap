// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination 组件快照 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "center",
      },
    ]
  }
>
  <View
    accessible={true}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        Object {
          "backgroundColor": "#fff",
          "borderColor": "#ff6644",
          "borderRadius": 4,
          "borderWidth": 1,
          "height": 42,
          "paddingHorizontal": 40,
        },
        Object {
          "height": 32,
        },
        Array [
          Object {
            "paddingHorizontal": 10,
          },
        ],
        null,
        Object {
          "backgroundColor": "#ccc",
          "borderColor": "#ccc",
        },
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#ff6644",
            "textAlign": "center",
          },
          Object {
            "fontSize": 13,
            "lineHeight": 30,
          },
          undefined,
          Object {
            "color": "#fff",
            "textAlign": "center",
          },
        ]
      }
    >
      上一页
    </Text>
  </View>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "flex": 1,
        },
      ]
    }
  >
    <Text>
      <Text
        style={
          Array [
            Object {
              "color": "#ff6644",
            },
          ]
        }
      >
        1
      </Text>
       / 
      20
    </Text>
  </View>
  <View
    accessible={true}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        Object {
          "backgroundColor": "#fff",
          "borderColor": "#ff6644",
          "borderRadius": 4,
          "borderWidth": 1,
          "height": 42,
          "paddingHorizontal": 40,
        },
        Object {
          "height": 32,
        },
        Array [
          Object {
            "paddingHorizontal": 10,
          },
        ],
        null,
        null,
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#ff6644",
            "textAlign": "center",
          },
          Object {
            "fontSize": 13,
            "lineHeight": 30,
          },
          undefined,
          null,
        ]
      }
    >
      下一页
    </Text>
  </View>
</View>
`;
